src/material/badge/badge.ts

```diff
@@ -225,56 +202,52 @@ export class MatBadge extends _MatBadgeBase implements OnDestroy, OnChanges, Can
     return badgeElement;
   }
 
-  /** Sets the aria-label property on the element */
-  private _updateHostAriaDescription(newDescription: string, oldDescription: string): void {
-    // ensure content available before setting label
-    const content = this._updateTextContent();
+  /** Update the text content of the badge.  */
+  private _updateContent(newContent: string | number | undefined | null): void {
+    const newContentNormalized: string = `${newContent ?? ''}`.trim();
+    if (newContentNormalized && !this._badgeElement) {
+      this._badgeElement = this._createBadgeElement();
+    }
 
-    if (oldDescription) {
-      this._ariaDescriber.removeDescription(content, oldDescription);
+    if (this._badgeElement) {
+      this._badgeElement.textContent = newContentNormalized;
     }
 
+    this._content = newContentNormalized;
+    this._hasContent = !!newContentNormalized;
+  }
+
+  /** Updates the host element's aria description via AriaDescriber. */
+  private _updateHostAriaDescription(newDescription: string): void {
+    this._ariaDescriber.removeDescription(this._elementRef.nativeElement, this.description);
     if (newDescription) {
-      this._ariaDescriber.describe(content, newDescription);
+      this._ariaDescriber.describe(this._elementRef.nativeElement, newDescription);
     }
+    this._description = newDescription;
   }
 
   /** Adds css theme class given the color to the component host */
   private _setColor(colorPalette: ThemePalette) {
-    if (colorPalette !== this._color) {
-      const classList = this._elementRef.nativeElement.classList;
-      if (this._color) {
-        classList.remove(`mat-badge-${this._color}`);
-      }
-      if (colorPalette) {
-        classList.add(`mat-badge-${colorPalette}`);
-      }
+    const classList = this._elementRef.nativeElement.classList;
+    classList.remove(`mat-badge-${this._color}`);
```

Comment:
Not much of a perf hit, but I think that technically this will call `.remove('mat-badge-undefined')` if the `_color` isn't defined.

Reply:
Yeah, I figured `color` rarely changes, so it's fine to hit the no-op `remove` here